compiler/rustc_resolve/src/late/diagnostics.rs

```diff
@@ -2605,6 +2587,67 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
     }
 }
 
+fn mk_where_bound_predicate(
+    path: &Path,
+    poly_trait_ref: &ast::PolyTraitRef,
+    ty: &ast::Ty,
+) -> Option<ast::WhereBoundPredicate> {
+    use rustc_span::DUMMY_SP;
+    let modified_segments = {
+        let mut segments = path.segments.clone();
+        let [preceding @ .., second_last, last] = segments.as_mut_slice() else { return None; };
+        let mut segments = ThinVec::from(preceding);
+
+        let added_constraint = ast::AngleBracketedArg::Constraint(ast::AssocConstraint {
+            id: DUMMY_NODE_ID,
+            ident: last.ident,
+            gen_args: None,
+            kind: ast::AssocConstraintKind::Equality {
+                term: ast::Term::Ty(ast::ptr::P(ast::Ty {
+                    kind: ast::TyKind::Path(None, poly_trait_ref.trait_ref.path.clone()),
+                    id: DUMMY_NODE_ID,
+                    span: DUMMY_SP,
+                    tokens: None,
+                })),
+            },
+            span: DUMMY_SP,
+        });
+
+        if let Some(ast::GenericArgs::AngleBracketed(ast::AngleBracketedArgs { args, .. })) =
+            second_last.args.as_deref_mut()
+        {
+            args.push(added_constraint);
+        } else {
+            second_last.args =
+                Some(ast::ptr::P(ast::GenericArgs::AngleBracketed(ast::AngleBracketedArgs {
+                    args: ThinVec::from([added_constraint]),
+                    span: DUMMY_SP,
+                })));
+        }
```

Comment:
I will add a check here just to be sure

```rust
if second_last.args.as_ref().is_some_and(|a| !a.is_angle_bracketed()) {
    return None;
}
```

I want to say you're right it is possible, maybe something like `Fn(A, B)`? I'm not 100% sure on when it can happen, but it probably can, so I added a check to bail if it's not angle bracketed.

Reply:
Please merge the two if conditions and the else branch into a single large match. The logic you have is correct, but I think it's more readable if all three variants are individual match arms